MaxText/layers/attentions.py

```diff
@@ -490,61 +506,92 @@ def decode(
     """
-    # Detect if we're initializing by absence of existing cache data.
-    is_initialized = self.has_variable('cache', 'cached_key')
-
-    cached_key = self.variable('cache', 'cached_key', jnp.zeros,
-                               self.cached_kv_shape(key.shape), key.dtype)
-    cached_value = self.variable('cache', 'cached_value', jnp.zeros,
-                                 self.cached_kv_shape(value.shape), value.dtype)
-    cache_index = self.variable('cache', 'cache_index',
-                                lambda: jnp.array(0, dtype=jnp.int32))
-
-    if not is_initialized:
-      return key, value
-
-    batch, num_heads, head_dim, length = cached_key.value.shape
-    # During fast autoregressive decoding, we feed one position at a time,
-    # and cache the keys and values step by step.
-    # Sanity shape check of cached key against input query.
-    expected_shape = (batch, 1, num_heads, head_dim)
-    if expected_shape != query_shape:
-      raise ValueError(f"""Autoregressive cache shape error,
-                        expected query shape %s instead got
-                        {(expected_shape, query_shape)}""")
-    # Create a OHE of the current index. NOTE: the index is increased below.
-    cur_index = cache_index.value
-    one_hot_indices = jax.nn.one_hot(cur_index, length, dtype=key.dtype)
-
-    # In order to update the key, value caches with the current key and
-    # value, we move the length axis to the back, similar to what we did for
-    # the cached ones above.
-    # Note these are currently the key and value of a single position, since
-    # we feed one position at a time.
-    one_token_key = self.move_kvlen_axis(key)
-    one_token_value = self.move_kvlen_axis(value)
-
-    # Update key, value caches with our new 1d spatial slices.
-    # We implement an efficient scatter into the cache via one-hot
-    # broadcast and addition.
-    key = cached_key.value + one_token_key * one_hot_indices
-    value = cached_value.value + one_token_value * one_hot_indices
-    cached_key.value = key
-    cached_value.value = value
-    cache_index.value = cache_index.value + 1
-
-    # Move the keys and values back to their original shapes.
-    key = self.revert_kvlen_axis(key)
-    value = self.revert_kvlen_axis(value)
-
-    return key, value
+    cache_logical_shape = (key.shape[0], self.max_target_length, key.shape[2], key.shape[3])
+
+
+    if key.shape != value.shape:
+      raise ValueError(f"Can't KV cache with mismatched shapes {key.shape=}, {value.shape=}")
+
+    if model_mode == common_types.TRAIN_MODEL_MODE:
+      return key, value, decoder_segment_ids
+    elif model_mode == common_types.PREFILL_MODEL_MODE:
+      batch, sequence, num_heads, head_dim = key.shape
+
+      cached_key = self.variable('cache', 'cached_key', jnp.zeros,
+                                self.cached_kv_shape(cache_logical_shape), key.dtype)
+      cached_key.value *= 0 # we might be prefilling something that already exists!
```

Comment:
this is just in case we're prefilling with a preexising cache. Then the above line just does a load (no `jnp.zeros`).

(I.e. the user is prefilling a new set of sequnces into the same fixed size cache.)